aten/src/ATen/native/mkldnn/RNN.cpp

```diff
@@ -0,0 +1,611 @@
+std::vector<int64_t> _output_size(const RNNParams& rnn) {
+  auto output_channels = is_single_direction ? rnn.hidden_size
+                                             : rnn.hidden_size * rnn.num_directions;
+  return {rnn.seq_length, rnn.mini_batch, output_channels};
+}
+
+// MKLDNN GRU gate order is different from PyTorch's which requires gates shuffle
+// (let rt,zt,nt be reset, update, new gates respectively)
+//
+//   MKLDNN GRU weight_ih/weight_hh gates order: (zt, rt, nt)
+//   PyTorch GRU weight_ih/weight_hh gates order: (rt, zt, nt)
+//
+// MKLDNN GRU bias has 4 gates instead of 3
+//  (PyTorch GRU bias)     (MKLDNN GRU bias)
+//
+//  bias_ih    bias_hh          bias
+//  +-----+    +-----+       +---------+
+//  | rt1 |    | rt2 |       | zt1+zt2 |
+//  |-----|    |-----|       |---------|
+//  | zt1 |    | zt2 |       | rt1+rt2 |
+//  |-----|    |-----|       |---------|
+//  | nt1 |    | nt2 |       |   nt1   |
+//  +-----+    +-----+       |---------|
+//                           |   nt2   |
+//                           +---------+
+//
+Tensor _shuffle_weight(const Tensor& weight, int64_t fn_mode) {
+  auto weight_t = weight.contiguous();
+  if (static_cast<ideep::rnn_kind>(fn_mode) == ideep::rnn_kind::GRU) {
+    std::vector<Tensor> gates = weight_t.chunk(3, /*gates*/0);
+    return at::cat({gates[1], gates[0], gates[2]}, /*gates*/0);
+  }
+  return weight_t;
+};
+
+Tensor _shuffle_bias(const Tensor& bias_ih, const Tensor& bias_hh, int64_t fn_mode) {
+  if (static_cast<ideep::rnn_kind>(fn_mode) == ideep::rnn_kind::GRU) {
+    std::vector<Tensor> b1 = bias_ih.chunk(3, /*output_channels*/0);
+    std::vector<Tensor> b2 = bias_hh.chunk(3, /*output_channels*/0);
+    return at::cat({b1[1] + b2[1], b1[0] + b2[0], b1[2], b2[2]}, /*output_channels*/0);
+  }
+  return bias_ih + bias_hh;
+};
+
+// Create mkldnn memory view from ATen tensor
+static inline ideep::tensor get_mkldnn_tensor(
+    const Tensor& tensor, const ideep::tensor::desc& desc) {
+  TORCH_CHECK(
+      tensor.device().is_cpu(),
+      "get_mkldnn_tensor expects CPU tensor input");
+  TORCH_CHECK(
+      tensor.layout() == at::Layout::Strided,
+      "get_mkldnn_tensor expects dense tensor input");
+  TORCH_CHECK(
+      tensor.scalar_type() == at::ScalarType::Float ||
+          tensor.scalar_type() == at::ScalarType::BFloat16 ||
+          tensor.scalar_type() == at::ScalarType::Half,
+      "get_mkldnn_tensor expects float or bfloat16 tensor input");
+  return {desc, tensor.data_ptr()};
+}
+
+// Init a aten tensor according to ideep tensor's desc.
+static inline Tensor empty_aten_tensor_from_desc(
+    const ideep::tensor::desc& desc,
+    const at::TensorOptions& options) {
+  auto ndims = desc.data.ndims;
+  auto nblks = desc.blocking_desc().inner_nblks;
+  std::vector<int64_t> at_sizes(ndims + nblks);
+  auto padded_dims = desc.padded_dims();
+  auto blk_sizes = desc.blocking_desc().inner_blks;
+  auto blk_idxs = desc.blocking_desc().inner_idxs;
```

Comment:
Do we really have to parse the internal blocking descriptors of onednn to get the workspace aten tensor? Can we just model it as a 1D tensor buffer from aten side?